apps/worker/src/transaction_notifier/notifier.service.ts

```diff
@@ -1,93 +1,161 @@
-          } else {
-            throw new Error(`Job ${job.data.id} failed with error ${JSON.stringify(txResult.error)}`);
-          }
-        }
+      const savedCapacity = await this.cacheManager.get(epochCapacityKey);
+      const epochCapacity = BigInt(savedCapacity ?? 0);
+      const newEpochCapacity = epochCapacity + capacityWithdrawn;
+
+      const epochDurationBlocks = await this.blockchainService.getCurrentEpochLength();
+      const epochDuration = epochDurationBlocks * SECONDS_PER_BLOCK * MILLISECONDS_PER_SECOND;
+      await this.cacheManager.setex(epochCapacityKey, epochDuration, newEpochCapacity.toString());
+    } catch (error) {
+      this.logger.error(`Error setting epoch capacity: ${error}`);
+    }
+  }
+
+  protected async checkInitialScanParameters(): Promise<boolean> {
+    const pendingTxns = await this.cacheManager.hlen(RedisUtils.TXN_WATCH_LIST_KEY);
+    return pendingTxns > 0;
+  }
+
+  protected async checkScanParameters(): Promise<boolean> {
+    const pendingTxns = await this.cacheManager.hlen(RedisUtils.TXN_WATCH_LIST_KEY);
+    return pendingTxns > 0;
+  }
+
+  public async getLastSeenBlockNumber(): Promise<number> {
+    let blockNumber = await super.getLastSeenBlockNumber();
+    const pendingTxns = await this.cacheManager.hvals(RedisUtils.TXN_WATCH_LIST_KEY);
+    if (pendingTxns.length > 0) {
+      const startingBlock = Math.min(
+        ...pendingTxns.map((valStr) => {
+          const val = JSON.parse(valStr) as ITxStatus;
+          return val.birth;
+        }),
+      );
+
+      blockNumber = Math.max(blockNumber, startingBlock);
+    }
+
+    return blockNumber;
+  }
 
-        if (txResult.success) {
-          this.logger.verbose(`Successfully found ${job.data.txHash} found in block ${txResult.blockHash}`);
+  async processCurrentBlock(currentBlockHash: BlockHash, currentBlockNumber: number): Promise<void> {
+    // Get set of tx hashes to monitor from cache
+    const pendingTxns = (await this.cacheManager.hvals(RedisUtils.TXN_WATCH_LIST_KEY)).map(
+      (val) => JSON.parse(val) as ITxStatus,
+    );
+
+    const block = await this.blockchainService.getBlock(currentBlockHash);
+    const extrinsicIndices: [HexString, number][] = [];
+    block.block.extrinsics.forEach((extrinsic, index) => {
+      if (pendingTxns.some(({ txHash }) => txHash === extrinsic.hash.toHex())) {
+        extrinsicIndices.push([extrinsic.hash.toHex(), index]);
+      }
+    });
+
+    if (extrinsicIndices.length > 0) {
+      const at = await this.blockchainService.api.at(currentBlockHash);
+      const epoch = (await at.query.capacity.currentEpoch()).toNumber();
+      const events: FrameSystemEventRecord[] = (await at.query.system.events()).filter(
+        ({ phase }) => phase.isApplyExtrinsic && extrinsicIndices.some((index) => phase.asApplyExtrinsic.eq(index)),
+      );
+
+      const totalCapacityWithdrawn: bigint = events
+        .filter(({ event }) => at.events.capacity.CapacityWithdrawn.is(event))
```

Comment:
This needs to filter down to just the provider that this is scanning for. Without it, a second provider's usage would be counted.

Reply:
We've already filtered down to events ONLY from the extrinsics in the current block that we submitted. That was part of a bug that this fixes; previously, we were looking at ALL events in the block.

Reply:
Ah I see it now. I missed it the first time around: `extrinsicIndices.some()`